src/shogun/util/visitors/InterfaceTypeVisitor.h

```diff
@@ -0,0 +1,74 @@
+#ifndef _BASE_TYPE_VISITOR_H_
+#define _BASE_TYPE_VISITOR_H_
+
+#include <shogun/base/base_types.h>
+
+#include <memory>
+
+namespace shogun
+{
+	class SGObject;
+
+	class InterfaceTypeVisitor
+	{
+	public:
+		virtual ~InterfaceTypeVisitor() = default;
+
+		virtual void on(std::shared_ptr<SGObject>*) = 0;
+		virtual void on(std::shared_ptr<Machine>*) = 0;
+		virtual void on(std::shared_ptr<Kernel>*) = 0;
+		virtual void on(std::shared_ptr<Distance>*) = 0;
+		virtual void on(std::shared_ptr<Features>*) = 0;
+		virtual void on(std::shared_ptr<Labels>*) = 0;
+		virtual void on(std::shared_ptr<ECOCEncoder>*) = 0;
+		virtual void on(std::shared_ptr<ECOCDecoder>*) = 0;
+		virtual void on(std::shared_ptr<Evaluation>*) = 0;
+		virtual void on(std::shared_ptr<EvaluationResult>*) = 0;
+		virtual void on(std::shared_ptr<MulticlassStrategy>*) = 0;
+		virtual void on(std::shared_ptr<NeuralLayer>*) = 0;
+		virtual void on(std::shared_ptr<SplittingStrategy>*) = 0;
+		virtual void on(std::shared_ptr<Pipeline>*) = 0;
+		virtual void on(std::shared_ptr<SVM>*) = 0;
+		virtual void on(std::shared_ptr<LikelihoodModel>*) = 0;
+		virtual void on(std::shared_ptr<MeanFunction>*) = 0;
+		virtual void on(std::shared_ptr<DifferentiableFunction>*) = 0;
+		virtual void on(std::shared_ptr<Inference>*) = 0;
+		virtual void on(std::shared_ptr<LossFunction>*) = 0;
+		virtual void on(std::shared_ptr<Tokenizer>*) = 0;
+		virtual void on(std::shared_ptr<CombinationRule>*) = 0;
+		virtual void on(std::shared_ptr<KernelNormalizer>*) = 0;
+		virtual void on(std::shared_ptr<Transformer>*) = 0;
+		virtual void on(std::shared_ptr<MachineEvaluation>*) = 0;
+		virtual void on(std::shared_ptr<StructuredModel>*) = 0;
+		virtual void on(std::shared_ptr<FactorType>*) = 0;
+		virtual void on(std::shared_ptr<ParameterObserver>*) = 0;
+		virtual void on(std::shared_ptr<Distribution>*) = 0;
+		virtual void on(std::shared_ptr<GaussianProcess>*) = 0;
+		virtual void on(std::shared_ptr<Alphabet>*) = 0;
```

Comment:
I wonder if there could be a way to automatically generate this? Then again the base types probably do not change that often

Reply:
We can do type erasure as follows
```C++
virtual void on(std::shared_ptr<void>*) = 0;
```
and then try to `dynamic_cast` for all base types one by one and dispatch to a generic `on`.
```C++
template <typename BaseClass>
void on(std::shared_ptr<BaseClass>*);
```

Reply:
We can manually construct the virtual table of `on` methods using a polymorphic lambda. So we would have something like
```C++
std::tuple<std::function<void(std::shared_ptr<BaseClasses>*)>...> vtable;

template <typename BaseClass>
void on(std::shared_ptr<BaseClass>* ptr)
{
   std::get<TypeToIndex<BaseClass>::value>(this->vtable) (ptr);
}
```
This should be just as efficient, but it requires complex code.

Reply:
I think the manual vtable is similar to what we have in Any and use in SGObject
https://github.com/shogun-toolbox/shogun/blob/9b8d856971af5a295dd6ad70623ae45647a6334c/src/shogun/base/SGObject.h#L768
I wonder if the same logic could be used. I guess it would only be possible if we add another base class between SGObject and the "base" classes, where it would register types to be used (somehow) in the interfaces

Reply:
The type erasure solution is kind of what we have/had in SGObject but we are moving away from that and just use visitor patterns.